File: server/src/services/auth-service.ts
import client from '../database';
import { BadRequestError } from '../exceptions/bad-request-exception';
import { Exception } from '../exceptions/exception';
import { InternalServerError } from '../exceptions/internal-server-exception';
import {
    findByEmailQuery,
    loginQuery,
    registerQuery,
    verifyUserQuery,
} from '../queries/auth-queries';
import { Helper } from '../util/helper';

export class AuthService {
    private helper: Helper;

    constructor() {
        this.helper = new Helper();
    }

    public async login(email: string, password: string) {
        const hashedPassword = this.helper.hashPassword(password);
        const result = await client.query(loginQuery, [email, hashedPassword]);
        if (result.rows.length === 0) {
            throw new BadRequestError('User not found!');
        }
        const user = result.rows[0];
        const payload = {
            id: user.id,
            email: user.email,
        };

        const accessToken = this.helper.generateAccessToken(payload);
        const refreshToken = this.helper.generateRefreshToken(payload);
        return {
            accessToken,
            refreshToken,
        };
    }

    public async register(payload: { username: string; email: string; password: string }) {
        const hashedPassword = this.helper.hashPassword(payload.password);
        try {
            await client.query('BEGIN');
            const isEmailExist = await client.query(findByEmailQuery, [payload.email]);
            if (isEmailExist.rows.length > 0) {
                throw new BadRequestError('Email already exist!');
            } else {
                const newUser = await client.query(registerQuery, [
                    payload.username,
                    payload.email,
                    hashedPassword,
                ]);
                await client.query('COMMIT');
                return {
                    id: newUser.rows[0].id,
                    username: newUser.rows[0].username,
                    email: newUser.rows[0].email,
                };
            }
        } catch (error) {
            await client.query('ROLLBACK');
            if (error instanceof Exception) {
                throw error;
            } else {
                throw new InternalServerError('Something went wrong!');
            }
        }
    }

    public async verify(token: string) {
        try {
            const decodedToken: any = this.helper.decodeToken(token);
            const result = await client.query(verifyUserQuery, [decodedToken.id]);
            if (result.rows.length === 0) {
                throw new BadRequestError('User not found!');
            }
            const user = result.rows[0];
            return user;
        } catch (error) {
            if (error instanceof Exception) {
                throw error;
            } else {
                throw new InternalServerError('Something went wrong!');
            }
        }
    }
}

File: server/service/auth_service.py
from ..model import User
from werkzeug.exceptions import HTTPException, abort
from ..repository.user_repository import UserRepository
from ..util.helper import Helper


class AuthService:
    def __init__(self, user_repository: UserRepository):
        self.user_repository = user_repository

    def register(self, username: str, email: str, password: str):
            existing_user = self.user_repository.get_by_email(email)
            if existing_user:
                abort(400, description="User already exists!")
            password_hash = User.hash_password(password)
            user = self.user_repository.create(username=username, email=email, password=password_hash)
            return user

    def login(self, email: str, password: str) -> dict:
            user = self.user_repository.get_by_email(email)
            if not user or not user.check_password(password):
                abort(400, description="Invalid credentials!")
            access_token = Helper.generate_access_token({"id": user.id})
            refresh_token = Helper.generate_refresh_token({"id": user.id})
            return {"access_token": access_token, "refresh_token": refresh_token}



